In [ ]:
print('hello IE 507')

hello IE 507


In [ ]:
! pip install -q pyomo


EXERCISE 1

min 2x1 + 15x2 - 4x3 - 6x4 + 7x5 - 9x6 - 1.6x7

s.t.

x1 + x2 + x3 + x4 + x5 + x6 + x7 =200 

x1 + x2 >= 10

4 x3 + 2 x6 + 3 x7 <= 150

10 x1 - x6 + 3 x5 >= 20

x3 - x5 + 3 x7 >= 10

x1 >= 0  
   x2 >= 12   
x3>= -10    
x4 >= 8     
x5 >= 9     
x6 >= -1   
x7 >=0






In [ ]:
from pyomo.environ import*

In [ ]:
#create a model
model=ConcreteModel()

In [ ]:
#declare decision variables
model.x1=Var(domain=Reals)
model.x2=Var(domain=Reals)
model.x3=Var(domain=Reals)
model.x4=Var(domain=Reals)
model.x5=Var(domain=Reals)
model.x6=Var(domain=Reals)
model.x7=Var(domain=Reals)

In [ ]:
#declare Objective
model.cost = Objective(expr = 2*model.x1 + 15*model.x2 - 4*model.x3 - 6*model.x4 + 7*model.x5 - 9*model.x6 - 1.6*model.x7, sense=minimize)

In [ ]:
#declare constraints
model.con1=Constraint(expr= model.x1 + model.x2 + model.x3 + model.x4 + model.x5 + model.x6 + model.x7 == 200 )
model.con2=Constraint(expr= model.x1 + model.x2 >=10 )
model.con3=Constraint(expr= 4*model.x3 +2*model.x6 + 3*model.x7 <= 150  )
model.con4=Constraint(expr= 10*model.x1 - model.x6 + 3*model.x5 >=20 )
model.con5=Constraint(expr= model.x3 - model.x5 + 3*model.x7 >=10 )
model.con6=Constraint(expr= model.x1 >= 0 )
model.con7=Constraint(expr= model.x2 >= 12 )
model.con8=Constraint(expr= model.x3 >= -10 )
model.con9=Constraint(expr= model.x4 >= 8 )
model.con10=Constraint(expr= model.x5 >= 9 )
model.con11=Constraint(expr= model.x6 >= -1 )
model.con12=Constraint(expr= model.x7 >= 0 )

In [ ]:
#printing the model
model.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x7 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fi

In [34]:
!apt-get install -y -qq glpk-utils

In [35]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -991.166666666667
  Upper bound: -991.166666666667
  Number of objectives: 1
  Number of constraints: 14
  Number of variables: 8
  Number of nonzeros: 27
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.011199951171875
# ----------------------------------------------------------
#   Solution Information
# -------------

In [38]:
# display solution
print('\nCost = ', model.cost())

print('\nDecision Variables')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())
print('x6 = ', model.x6())
print('x7 = ', model.x7()) 

print('\nConstraints')
print('Constraint 1 =',model.con1())
print('Constraint 2 =',model.con2())
print('Constraint 3 =',model.con3())
print('Constraint 4 =',model.con4())
print('Constraint 5 =',model.con5())
print('Constraint 6 =',model.con6())
print('Constraint 7 =',model.con7())
print('Constraint 8 =',model.con8())
print('Constraint 9 =',model.con9())
print('Constraint 10 =',model.con10())
print('Constraint 11=',model.con11())
print('Constraint 12=',model.con12())




Cost =  -991.1666666666665

Decision Variables
x1 =  7.35
x2 =  12.0
x3 =  -10.0
x4 =  91.4833333333333
x5 =  9.0
x6 =  80.5
x7 =  9.66666666666667

Constraints
Constraint 1 = 199.99999999999997
Constraint 2 = 19.35
Constraint 3 = 150.0
Constraint 4 = 20.0
Constraint 5 = 10.000000000000007
Constraint 6 = 7.35
Constraint 7 = 12.0
Constraint 8 = -10.0
Constraint 9 = 91.4833333333333
Constraint 10 = 9.0
Constraint 11= 80.5
Constraint 12= 9.66666666666667


(3) Report the objective function value , the value of all variables at the optimal point. 
 
 The objective function value = -991.1666666666665

 The value of all variablesis   
 x1 =  7.35   
x2 =  12.0     
x3 =  -10.0        
x4 =  91.4833333333333      
x5 =  9.0      
x6 =  80.5       
x7 =  9.66666666666667  

(4)  Report the constraint activities (the value of left hand side(LHS) in a constraint and if the constraint is satisfied)for first four constraints.

Constraint 1 = 199.99999999999997    
Constraint 2 = 19.35    
Constraint 3 = 150.0     
Constraint 4 = 20.0      

